### PR TITLE
Arelle local viewer updates

### PR DESCRIPTION
**Pull request by hefischer**
Updates for Arelle operation in local viewer with GUI and compatibility with EdgarRenderer.  Updates to work with REST web service API returning zip stream results.

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
The items listed below may not capture all security relevant changes. Before providing a security review, be sure to review the entire PR for security impact.
<details>
<summary>(3) Security relevant changes were detected</summary>
<li>Watched keyword <code>Cache-Control</code> in <code>iXBRLViewerPlugin/LocalViewer.py</code> line(s) ['60'] added</li>
<li>Watched keyword <code>etree</code> in <code>iXBRLViewerPlugin/xhtmlserialize.py</code> line(s) ['41'] added</li>
<li>Watched keyword <code>etree</code> in <code>iXBRLViewerPlugin/xhtmlserialize.py</code> line(s) ['41'] removed</li>
</details>

## Action Items
* Request infosec review via Slack: #support-infosec
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/teams/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: iXBRLViewerPlugin/LocalViewer.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                        return static_file("arelle.ico", root=cntlr.imagesDir, mimetype='image/vnd.microsoft.icon')
+                    _report, _sep, _file = file.partition("/")
+                    if _report.isnumeric(): # in reportsFolder folder
+                        if _file == "ixbrlviewer.js":
+                            root = os.path.join(os.path.dirname(__file__), "viewer", "dist")
+                            return static_file(_file, root=root, mimetype='text/javascript')
+                        else:
+                            root = reportsFolders[int(_report)]
+                        # print("file {} root {}".format(_file, root))
+                        # print (os.path.join(reportsFolders[int(_report)], _file))
```
Comment (derekgengenbacher-wf) [inline iXBRLViewerPlugin/LocalViewer.py]: Do we need these commented out print lines?

File: iXBRLViewerPlugin/LocalViewer.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            
+            # start server
+            localserver = Bottle()
+            
+            localserver.route('/<file:path>', 'GET', getlocalfile)
+            localserver.route('<relpath:path>', 'GET', getlocalfile)
+            # start local server on the port on a separate thread
+            threading.Thread(target=localserver.run, 
+                             # The wsgi server part of cherrypy was split into a new 'cheroot'
+                             #kwargs=dict(server='cherrypy', host='localhost', port=port, quiet=True),
```
Comment (derekgengenbacher-wf) [inline iXBRLViewerPlugin/LocalViewer.py]: Is this needed?

File: iXBRLViewerPlugin/LocalViewer.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                             # The wsgi server part of cherrypy was split into a new 'cheroot'
+                             #kwargs=dict(server='cherrypy', host='localhost', port=port, quiet=True), 
+                             kwargs=dict(server='cheroot', host='localhost', port=port, quiet=True), 
+                             daemon=True).start()
+            time.sleep(2) # allow other thread to run and start up
+    
+        localhost = "http://localhost:{}/{}".format(port, len(reportsFolders))
+        reportsFolders.append(reportsFolder)
+        cntlr.addToLog(_("iXBRLViewer local viewer: http://localhost:{}").format(port), messageCode="localViewer:listen",level=logging.DEBUG)
+        #cntlr.addToLog("localhost={}".format(localhost), messageCode="localViewer:listen",level=logging.DEBUG)
```
Comment (derekgengenbacher-wf) [inline iXBRLViewerPlugin/LocalViewer.py]: Is this needed?

File: iXBRLViewerPlugin/LocalViewer.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        if port is None: # already initialized
+        
+            # find available port
+            import socket
+            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            s.bind(("",0))
+            s.listen(1)
+            port = s.getsockname()[1]
+            s.close()
+            #port = 8080 # test with fixed port
```
Comment (derekgengenbacher-wf) [inline iXBRLViewerPlugin/LocalViewer.py]: Is this needed?

File: iXBRLViewerPlugin/LocalViewer.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            
+            localserver.route('/<file:path>', 'GET', getlocalfile)
+            localserver.route('<relpath:path>', 'GET', getlocalfile)
+            # start local server on the port on a separate thread
+            threading.Thread(target=localserver.run, 
+                             # The wsgi server part of cherrypy was split into a new 'cheroot'
+                             #kwargs=dict(server='cherrypy', host='localhost', port=port, quiet=True), 
+                             kwargs=dict(server='cheroot', host='localhost', port=port, quiet=True), 
+                             daemon=True).start()
+            time.sleep(2) # allow other thread to run and start up
```
Comment (derekgengenbacher-wf) [inline iXBRLViewerPlugin/LocalViewer.py]: What other thread are we waiting for here? And can we not use join() to stop progression until both threads are done?

File: iXBRLViewerPlugin/__init__.py [hunk 6]
```diff
@@ -11,9 +11,33 @@
...
+#              -o ~/temp/out.zip 
+#              "http://localhost:8080/rest/xbrl/validation?&media=zip&plugins=iXBRLViewerPlugin&packages=somewhere/esef_taxonomy_2017.zip"
+#
+
 
 from .iXBRLViewer import IXBRLViewerBuilder
+import os, zipfile, sys, traceback
+from arelle.FileSource import archiveFilenameParts
 
+BUILTIN_SCRIPT_URL = os.path.join(os.path.dirname(__file__), "viewer", "dist", "ixbrlviewer.js")
```
Comment (derekgengenbacher-wf) [inline iXBRLViewerPlugin/__init__.py]: The root of `viewer/dist` is used in multiple places in this PR can we utilize this constant else where so that if its decided to move the ixbrlviewer.js file to a different loocation we dont have to update multiple places in the code?

File: iXBRLViewerPlugin/__init__.py [hunk 7]
```diff
@@ -23,7 +47,7 @@ def iXBRLViewerCommandLineOptionExtender(parser, *args, **kwargs):
     parser.add_option("--viewer-url",
                       action="store",
                       dest="viewerURL",
-                      default="js/dist/ixbrlviewer.js",
+                      # default="js/dist/ixbrlviewer.js",
```
Comment (derekgengenbacher-wf) [inline iXBRLViewerPlugin/__init__.py]: Is this needed?

File: iXBRLViewerPlugin/iXBRLViewer.py [hunk 8]
```diff
@@ -235,19 +237,14 @@ def createViewer(self, scriptUrl="js/dist/ixbrlviewer.js"):
 
             if f.context.isForeverPeriod:
                 aspects["p"] = "f"
-            elif f.context.isInstantPeriod and f.context.instantDatetime is not None:
+            elif f.context.isInstantPeriod:
                 aspects["p"] = self.dateFormat(f.context.instantDatetime.isoformat())
-            elif f.context.isStartEndPeriod and f.context.startDatetime is not None and f.context.endDatetime is not None:
```
Comment (derekgengenbacher-wf) [inline iXBRLViewerPlugin/iXBRLViewer.py]: Taking away both of these checks here seems slightly dangerous knowing that garbage data will run through here.

**Comment by derekgengenbacher-wf** [comment]
If I was going to QA this Im not sure what I would do to make sure that everything works correctly. Can you add some testing suggestions so that we can fully test this, please?

File: iXBRLViewerPlugin/LocalViewer.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Portions are from Edgar(tm) Renderer which was created by staff of the U.S. Securities and Exchange Commission.
+# Data and content created by government employees within the scope of their employment 
+# are not subject to domestic copyright protection. 17 U.S.C. 105.
```
Comment (derekgengenbacher-wf) [inline iXBRLViewerPlugin/LocalViewer.py]: Im not real sure how all this copyright stuff works. It seems a little weird that this file alone would be copyrighted by `Mark V Systems` and the rest of the repo by `Workiva`. Not suggesting we change it, but just calling it out as an inconsistency in the copyrighting.

File: iXBRLViewerPlugin/iXBRLViewer.py [hunk 10]
```diff
@@ -235,19 +237,14 @@ def createViewer(self, scriptUrl="js/dist/ixbrlviewer.js"):
 
             if f.context.isForeverPeriod:
                 aspects["p"] = "f"
-            elif f.context.isInstantPeriod and f.context.instantDatetime is not None:
+            elif f.context.isInstantPeriod:
```
Comment (brettkail-wk) [inline iXBRLViewerPlugin/iXBRLViewer.py]: Why this change (and the `isStartEndPeriod` below)?  This appears to revert the fixes from #70.

File: iXBRLViewerPlugin/iXBRLViewer.py (hunk 8 above)
Comment (brettkail-wk) [reply iXBRLViewerPlugin/iXBRLViewer.py]: Yes, I agree.  My GitHub view was stale, so I ended up making a similar comment [here](https://github.com/Workiva/ixbrl-viewer/pull/74#discussion_r365251531).

**Comment by hefischer** [comment]
Let’s have a discussion on LocalViewer.py.  
Since it is duplicated between the SEC ix viewer and iXBRL Viewer I think it should be moved to the infrastructure of Arelle (and parameterized so it can work for both viewers concurrently).
This may be a good time to review it’s functionality as I think a future direction for Arelle GUI is to move away from Tkinter to browser-local GUI, like pgAdmin4.
Some EDGAR and ESEF filers may wish to preview filings in both ix viewers.
The threading is related to how some browsers start up numerous sockets for each included graphical component.
Other modules in the pull request reflect code from the most recent merge last year, so there may be intermediate updates needed on my end.

    Herm

File: iXBRLViewerPlugin/__init__.py [hunk 11]
```diff
@@ -77,8 +154,12 @@ def commandLineRun(*args, **kwargs):
     'license': 'License :: OSI Approved :: Apache License, Version 2.0 (Apache-2.0)',
     'author': 'Paul Warren',
     'copyright': 'Copyright :: Workiva Inc. :: 2019',
-    'imports': ["./iXBRLViewer.py"],
+    #'imports': ["./iXBRLViewer.py"],
```
Comment (sagesmith-wf) [inline iXBRLViewerPlugin/__init__.py]: Delete

File: iXBRLViewerPlugin/iXBRLViewer.py [hunk 12]
```diff
@@ -235,19 +237,14 @@ def createViewer(self, scriptUrl="js/dist/ixbrlviewer.js"):
...
+            elif f.context.isStartEndPeriod:
                 aspects["p"] = "%s/%s" % (
                     self.dateFormat(f.context.startDatetime.isoformat()),
                     self.dateFormat(f.context.endDatetime.isoformat())
                 )
 
-            frels = self.footnoteRelationshipSet.fromModelObject(f)
-            if frels:
-                for frel in frels:
-                    factData.setdefault("fn", []).append(frel.toModelObject.id)
```
Comment (paulwarren-wk) [inline iXBRLViewerPlugin/iXBRLViewer.py]: This removes footnote support (reverts #72)

File: iXBRLViewerPlugin/iXBRLViewer.py [hunk 13]
```diff
@@ -154,26 +154,28 @@ def addConcept(self, concept):
...
+            if r.toModelObject is not None:
+                self.treeWalk(rels, r.toModelObject, indent + 1)
 
     def getRelationships(self):
         rels = {}
 
         for baseSetKey, baseSetModelLinks  in self.dts.baseSets.items():
             arcrole, ELR, linkqname, arcqname = baseSetKey
-            if arcrole == XbrlConst.summationItem and ELR is not None:
+            if (arcrole == XbrlConst.parentChild or arcrole == XbrlConst.summationItem) and ELR is not None:
```
Comment (paulwarren-wk) [inline iXBRLViewerPlugin/iXBRLViewer.py]: This is reverting #49.

File: iXBRLViewerPlugin/iXBRLViewer.py [hunk 14]
```diff
@@ -154,26 +154,28 @@ def addConcept(self, concept):
...
                 for r in relSet.modelRelationships:
-                    fromKey = self.nsmap.qname(r.fromModelObject.qname)
-                    rel = {
-                        "t": self.nsmap.qname(r.toModelObject.qname),
-                    }
-                    if r.weight is not None:
-                        rel['w'] = r.weight
-                    rr.setdefault(fromKey, []).append(rel)
-                    self.addConcept(r.toModelObject)
+                    if r.fromModelObject is not None and r.toModelObject is not None:
```
Comment (paulwarren-wk) [inline iXBRLViewerPlugin/iXBRLViewer.py]: I'm not sure how it is possible to get a relationship which doesn't have a from or to object in Arelle, but if it is, then this guard is a useful addition.

File: iXBRLViewerPlugin/iXBRLViewer.py [hunk 15]
```diff
@@ -309,21 +306,28 @@ def __init__(self, dts):
...
-                filename = os.path.join(outPath, f.filename)
-                self.dts.info("viewer:info", "Writing %s" % filename)
-                with open(filename, "wb") as fout:
-                    writer = XHTMLSerializer()
-                    writer.serialize(f.xmlDocument, fout)
+                self.dts.info("viewer:info", "Writing %s" % f.filename)
+                writer = XHTMLSerializer()
+                xhtmlStr = writer.serialize(f.xmlDocument)
+                if responseZip:
+                    responseZip.writestr(f.filename, xhtmlStr)
```
Comment (paulwarren-wk) [inline iXBRLViewerPlugin/iXBRLViewer.py]: I think this is the key change that's actually needed in this file, although it needs rewriting to use the new signature for writer.serialize which takes a stream (probably needs io.BytesIO)

File: iXBRLViewerPlugin/__init__.py [hunk 16]
```diff
@@ -32,11 +56,30 @@ def iXBRLViewerCommandLineXbrlRun(cntlr, options, *args, **kwargs):
...
-        viewerBuilder = IXBRLViewerBuilder(cntlr.modelManager.modelXbrl)
-        iv = viewerBuilder.createViewer(scriptUrl=options.viewerURL)
-        iv.save(out)
+    try:
+        modelXbrl = cntlr.modelManager.modelXbrl
+        responseZipStream = kwargs.get("responseZipStream")
+        responseZip = None
+        out = getattr(options, 'saveViewerFile', False)
+        if not out and responseZipStream:
+            out = modelXbrl.modelDocument.basename.rpartition(".")[0] + "-ixbrlView.html"
```
Comment (paulwarren-wk) [inline iXBRLViewerPlugin/__init__.py]: I don't think this will cope with inline XBRL document sets.

File: iXBRLViewerPlugin/LocalViewer.py (hunk 1 above)
Comment (hefischer) [reply iXBRLViewerPlugin/LocalViewer.py]: No, removed in refactored LocalViewer.py

File: iXBRLViewerPlugin/LocalViewer.py (hunk 2 above)
Comment (hefischer) [reply iXBRLViewerPlugin/LocalViewer.py]: Yes, but moved to arelle.LocalViewer in refactored module

File: iXBRLViewerPlugin/LocalViewer.py (hunk 3 above)
Comment (hefischer) [reply iXBRLViewerPlugin/LocalViewer.py]: Yes, but moved to arelle.LocalViewer in refactored module

File: iXBRLViewerPlugin/LocalViewer.py (hunk 4 above)
Comment (hefischer) [reply iXBRLViewerPlugin/LocalViewer.py]: Yes, but moved to arelle.LocalViewer in refactored module

File: iXBRLViewerPlugin/LocalViewer.py (hunk 5 above)
Comment (hefischer) [reply iXBRLViewerPlugin/LocalViewer.py]: there are numerous threads set up for the Browser to be able to retrieve over multiple sockets, needed to let them settle.

File: iXBRLViewerPlugin/__init__.py (hunk 6 above)
Comment (hefischer) [reply iXBRLViewerPlugin/__init__.py]: Of course

File: iXBRLViewerPlugin/iXBRLViewer.py (hunk 8 above)
Comment (hefischer) [reply iXBRLViewerPlugin/iXBRLViewer.py]: old copy, added these in

File: iXBRLViewerPlugin/LocalViewer.py (hunk 9 above)
Comment (hefischer) [reply iXBRLViewerPlugin/LocalViewer.py]: the Mark V stuff was refactored to infrastructure, now this module uses W's legalese

File: iXBRLViewerPlugin/iXBRLViewer.py (hunk 10 above)
Comment (hefischer) [reply iXBRLViewerPlugin/iXBRLViewer.py]: stale copy, put back in

File: iXBRLViewerPlugin/__init__.py (hunk 11 above)
Comment (hefischer) [reply iXBRLViewerPlugin/__init__.py]: Removed

File: iXBRLViewerPlugin/iXBRLViewer.py (hunk 13 above)
Comment (hefischer) [reply iXBRLViewerPlugin/iXBRLViewer.py]: fixed

**Comment by paulwarren-wk** [comment]
Herm has now moved LocalViewer.py into the Arelle code base.  I suggest we close this PR, and I'll work with Herm to create a fresh one that just has the necessary changes.
